### ISSUE TITLE
Spin-out builder

### ISSUE THREAD

**Issue by icecrime** [docker/docker#14298]
# Description

As indicated by our [roadmap](https://github.com/docker/docker/blob/master/ROADMAP.md), we want the builder to be pulled out of the Engine. The end goal is that the daemon should know nothing about the Dockerfile.

The lack of ability to copy files to a container makes it impossible to fully implement the builder using exclusively the API. We have ongoing PR for this, that would be a prerequisite for this PR:

- #13171: docker cp to and from containers
- #14058: more docker cp

Apparently @jlhawn also has work on his fork for a client-side builder.

# How to

- [x] Help review and merge the docker cp PR mentioned above
- [ ] Help review and see what pieces of @jlhawn can be integrated in the Engine

**Comment by jlhawn** [comment]
@docker/maintainers I've just rebased the PRs referenced above so they are back in a merge-able state after 2 weeks of falling behind the master branch.

So far, my client-side builder work has been happening in a separate project named [**dockramp**](https://github.com/jlhawn/dockramp) but it's more of an proof-of-concept for a client-side builder as well as new Dockerfile syntax. We definitely want to keep backwards compatibility if we spin out the builder to be client-side in Docker 1.x so we should continue to use the current Dockerfile parser and consider **dockramp** to be a potential `Dockerfile v2`.

**Comment by stevvooe** [comment]
:+1:

**Comment by icecrime** [comment]
Good news: #13171 is in. Let's try and work through the rest for 1.9.0.

**Comment by jlhawn** [comment]
Awesome. This means that `dockramp` will work with Docker 1.8 if people want to use it.

**Comment by thaJeztah** [comment]
is there a way to add dockeramp as an experimental feature?

thinking; perhaps we should have a plugin system for the client after all, so that users can add custom builders; `docker build --builder=dockramp -f Dockerfile.dockramp`?

**Comment by thedrow** [comment]
@jlhawn Can you please explain how are you running commands inside a container during build time and how do you implement caching for a client side builder?

**Comment by muayyad-alsadi** [comment]
@jlhawn  @thedrow  the way I think about it, is that we start a temporary container, do the build, commit the container into an image.

**Comment by thedrow** [comment]
@muayyad-alsadi Then how do you get the caching capabilities of the normal builder?
If you start a temporary docker container and build it using docker exec you won't get any caching.

**Comment by jlhawn** [comment]
@thedrow @muayyad-alsadi `COPY` and `RUN` creates a new container. For `COPY` files are copied into the container via the new API endpoint and the container is immediately committed having never been started. For `RUN` the container is started using the command specified by the `RUN`, then committed. A commit returns a new image ID.

Caching is done by the client storing a hash of the image ID + the build command(s) that would lead to the committed image. This hash maps to the committed image ID. The client stores this mapping in a local file. For every step of the build file, the builder does something like this:

```
I'm starting this build step with image ID `x` and about to execute a command that does `y`.
Is there a local mapping of `sha256(x+y)` -> image ID `z`?
yes? does it exist on the daemon?
yes? that's a cache hit and the build step can be skipped.
continue to the next build step with image ID `z`.
```

**Comment by blaggacao** [comment]
I see this in the 1.10 roadmap, but with little recent activity. Eyes are on this for better artifact management in combination with nested build steps for greater CI implementation flexibility. This is kind of the missing part to make things perfect at the far end of the CI Toolchain. /fyi @kurkop

**Comment by cpuguy83** [comment]
@blaggacao It's definitely being actively worked on.
Will it make 1.10? I'm not sure. Code freeze is coming

**Comment by icecrime** [comment]
A lot of work was achieved to decouple the current [`builder`](https://github.com/docker/docker/tree/master/builder) subsystems from other parts of the Engine (most notably from the daemon). As of Docker Engine 1.10, this sub-package could now be reused to provide out-of-daemon build capabilities using the API.

We won't go further on this effort ourselves, as backward compatibility constraint forces us to preserve a daemon-side implementation of the Builder anyway. I hope some people will catch up with this effort and propose interesting things outside of the daemon.

**Comment by thedrow** [comment]
So this won't be released separately?

**Comment by pjvleeuwen-oss** [comment]
So as a docker amateur I ran into the issue of not being able to set the correct `user:group` when doing a `COPY`. After spending quite of time trying to reverse engineer what happened on this issue I found #9934 by @thaJeztah et al that addresses this issue back in Jan 2015 (!). After quite some discussions it was decided to fix this with options which were actually implemented and merged May 2015 to master as part of #10775 @duglin @icecrime et al. But so far the `COPY` uid issue is still not fixed, but all preparations are done. Then May 2015 @kostickm tried to actually do the implementation that was prepared for and decided on in #13600. This was blocked Jul 2015 because Jun 2015 @icecrime started this effort to move builder logic to client side. End result is that after more then a year the `COPY` uid issue is still open (or better said: #9934 and #13600 are actually closed, but the actual problem is not resolved). So what is the plan on this? Are we now in a right state to reattempt #13600 ?
